### PR TITLE
[N3] Update types to match N3 1.16

### PR DESCRIPTION
**Pull request by mielvds**
This PR updates the types for https://github.com/rdfjs/N3.js to match the changes introduced in v1.16.x .

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## n3 ([<kbd>unpkg</kbd>](https://unpkg.com/browse/n3@latest/))
was missing the following properties:
1. Term
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 8cb5f51f4789ee175ea62ab132b83c082eecc2cc
</p>

File: types/n3/index.d.ts
Comment (jeswr) [inline types/n3/index.d.ts:196]: Whenever there is an `EventEmitter` a `ParseCallback` must be provided, otherwise there is a runtime error.

```suggestion
    parse(input: string, callback?: null | undefined, prefixCallback?: PrefixCallback): Q[];
    parse(input: string | EventEmitter, callback: ParseCallback<Q>, prefixCallback?: PrefixCallback): void;
```

File: types/n3/index.d.ts
Comment (jeswr) [inline types/n3/index.d.ts:252]:
```suggestion
```

The object of `readQuads` is cannot be an array AFAIK.

I also don't think we should be exposing this since `#match` now provides the same functionality and is a standard RDF/JS method.

File: types/n3/tsconfig.json
Comment (jeswr) [inline types/n3/tsconfig.json:8]: What is the reason for this?

**Review by jeswr — CHANGES_REQUESTED** [review]

File: types/n3/n3-tests.ts
Comment (jeswr) [inline types/n3/n3-tests.ts:161]: What is the reason for definining this type?

File: types/n3/n3-tests.ts
Comment (jeswr) [inline types/n3/n3-tests.ts:265]:
```suggestion
    const quadGenerator: Generator<RDF.Quad> = store.match(N3.DataFactory.namedNode('http://ex.org/Mickey'), null, null, null);
```

**Review by jeswr — CHANGES_REQUESTED** [review]

**Comment by jeswr** [comment]
Thanks for your work @mielvds - lmk if the requested changes make sense :)

As a follow up it might also be good if you could make a PR to the N3.js readme explaining that `#match` should be used in favour of `#readQuads` and add an `@deprecated` jsdoc tag to the method.

File: types/n3/n3-tests.ts
Comment (mielvds) [reply types/n3/n3-tests.ts:265]: But isn't this update supposed to include the existence of `readQuads`? Hence the test should test that?

Comment (jeswr) [reply types/n3/n3-tests.ts:265]: See https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65275#discussion_r1212515869

File: types/n3/index.d.ts
Comment (mielvds) [reply types/n3/index.d.ts:251]: I think its better to expose it, but mark it as deprecated. It is in the docs for a while now and people started using it.

Comment (jeswr) [reply types/n3/index.d.ts:251]: > It is in the docs for a while now

Yes, because it pre-ceded the `#match` method because this library predates the RDF/JS spec.

> and people started using it.

Those people are already coping with this method not being available in the types (by ts-ignoring, type-casing etc.); so I don't see how this change benefits those users.


Either way I'll give @RubenVerborgh final call on this.

File: types/n3/tsconfig.json
Comment (jeswr) [reply types/n3/tsconfig.json:8]: @mielvds ?

File: types/n3/n3-tests.ts
Comment (jeswr) [reply types/n3/n3-tests.ts:161]: @mielvds ?

Comment (mielvds) [reply types/n3/n3-tests.ts:161]: I don't know, it was borrowed from a previous test: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/n3/n3-tests.ts#L131

Comment (mielvds) [reply types/n3/n3-tests.ts:161]: I don't know, it was borrowed from a previous test: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/n3/n3-tests.ts#L131

File: types/n3/index.d.ts
Comment (mielvds) [reply types/n3/index.d.ts:251]: But then it should be removed from N3.js first. I'm fine either way!

File: types/n3/tsconfig.json
Comment (mielvds) [reply types/n3/tsconfig.json:8]: I wasn't able to run the tests without

Comment (jeswr) [reply types/n3/tsconfig.json:8]: Which change required it?

The test suite should be passing with the original config before any of the changes in this PR were made - unless there have been changes to the DT setup that I don't know about.

File: types/n3/index.d.ts
Comment (jeswr) [inline types/n3/index.d.ts:194]: @mielvds can you add a test case where `undefined` is explicitly provided as a parameter?

Comment (jeswr) [reply types/n3/index.d.ts:194]: And I think we should keep the explicit `undefined` given the introduction of `--exactOptionalPropertyTypes` in TS 4.4 (https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-4.html)

Comment (mielvds) [reply types/n3/index.d.ts:194]: I removed the explicit `undefined` because the DefinitelyTyped tests were complaining: https://github.com/DefinitelyTyped/DefinitelyTyped/commit/c01e427dde51eaf526ed62ecc7c839119c2c551b/checks?check_suite_id=13299824905#step6:130. I added the tests

File: types/n3/tsconfig.json
Comment (mielvds) [reply types/n3/tsconfig.json]: Right... I got `ERROR: 270:21  expect  TypeScript@5.1 compile error: 
Type 'Generator<Quad, any, unknown>' can only be iterated through when using the '--downlevelIteration' flag or with a '--target' of 'es2015' or higher.` without

File: types/n3/index.d.ts
Comment (RubenVerborgh) [reply types/n3/index.d.ts:251]: I don't have any strong attachment to `readQuads`. The convention in N3.js is that, anything that does not start with an underscore, is part of the public API.

I would document for now, and we can indeed consider deprecation.

File: types/n3/index.d.ts
Comment (jeswr) [reply types/n3/index.d.ts:194]: Ah - my bad; it must only apply to properties on interfaces, rather than all optional types - thanks for adding the tests.

**Comment by jeswr** [comment]
Happy to approve once the build error is fixed (my preference would be using the type `Iterable<OutQuad>` instead of `Generator`).

Can we also do the following (perhaps in a follow up PR):
 - Import from `@rdfjs/types` rather than `rdf-js`
 - Use `RDF.Term` rather than `Term` (which is more strict as it has the `#toJSON` method) for parameters on methods like `#match`

File: types/n3/index.d.ts
Comment (mielvds) [reply types/n3/index.d.ts:251]: @jeswr while I agree with you, it's more of an architectural discussion of N3js and I don't think a PR about typings is the place to make those decisions :) Ok if we continue this discussion in the N3js repo in context of a next release? Happy to adjust the typings for that one

File: types/n3/tsconfig.json
Comment (mielvds) [reply types/n3/tsconfig.json]: @jeswr @RubenVerborgh not sure how to solve this one. Suggestions?

File: types/n3/index.d.ts
Comment (mielvds) [reply types/n3/index.d.ts:194]: So should I still make any changes or is it ok?

Comment (jeswr) [reply types/n3/index.d.ts:194]: It is ok

File: types/n3/tsconfig.json
Comment (jeswr) [reply types/n3/tsconfig.json:8]: Per https://github.com/DefinitelyTyped/DefinitelyTyped/pull/65275#issuecomment-1614124038 I would just make the return type `Iterable<OutQuad>`

Comment (mielvds) [reply types/n3/tsconfig.json:8]: done in 50ac2831d2ba52f3fc2c6401bdefb0ecca3e03cb

**Review by jeswr — APPROVED** [review]

File: types/n3/index.d.ts
Comment (jeswr) [reply types/n3/index.d.ts]: Yep; I suggest we mark it as deprecated in the next release https://github.com/rdfjs/N3.js/issues/355

**Review by jeswr — APPROVED** [review]

**Review by navya9singh — APPROVED** [review]

**Comment by jeswr** [comment]
Ready to merge

**Pull request merged**
